### PR TITLE
IDE-24702: Replacing QVCI in dbt-teradata

### PR DESCRIPTION
**Pull request by tallamohan**
QVCI is not advisable with teradata machines, dbt-teradata uses the QVCI feature to retrieve the view's metadata from the database.
The DBS and GSO teams advise not to recommend this to customers as there have been DBS restarts and crashes with QVCI-enabled systems.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -343,31 +343,6 @@ All dbt commands are supported.
...
-
-      # restart Teradata
-      tpareset -y Enable QVCI
-      ```
-  2. Instruct `dbt` to use `QVCI` mode. Include the following variable in your `dbt_project.yml`:
-      ```yaml
-      vars:
-        use_qvci: true
-      ```
-      For example configuration, see `test/catalog/with_qvci/dbt_project.yml`.
```
Comment (VarunSharma15) [inline README.md:369]: There is a section in this file at the end regarding Unit Testing, which talks about QVCI
Can we please remove that section as well ?!

File: dbt/include/teradata/macros/adapters.sql [hunk 2]
```diff
@@ -118,8 +118,27 @@
...
+        SELECT TableKind FROM DBC.TablesV WHERE DatabaseName = '{{ relation.schema }}' AND TableName = '{{ relation.identifier }}'
+    {% endcall %}
+
+    {% set table_kind = load_result('check_table_or_view').table.columns['TableKind'].values()[0] %}
+    {{ log("Table Kind : " ~ table_kind) }}
+
+    {%- if table_kind == 'V ' -%}
+        {{ log("Table Kind is V") }}
+        {% set temp_relation_for_view = relation.identifier ~ '_viw_tbl' %}
+        {{ log("Table temp_relation_for_view : " ~ temp_relation_for_view) }}
```
Comment (VarunSharma15) [inline dbt/include/teradata/macros/adapters.sql:131]: Is this logging really required ?
This might be confusing for end user, they don't have anything to do with this.

File: dbt/include/teradata/macros/adapters.sql [hunk 3]
```diff
@@ -118,8 +118,27 @@
...
+        {% call statement('drop_existing_table', fetch_result=False) %}
+            DROP table /*+ IF EXISTS */ "{{ relation.schema }}"."{{ temp_relation_for_view }}";
+        {% endcall %}
+        load_result('drop_existing_table')
+        {% call statement('creating_table_from_view', fetch_result=False) %}
+            CREATE TABLE "{{ relation.schema }}"."{{ temp_relation_for_view }}" AS (SELECT * FROM "{{ relation.schema }}"."{{ relation.identifier }}") WITH NO DATA;
+        {% endcall %}
+        load_result('creating_table_from_view')
+    {% endif %}
+    {{ log("Table Kind is not V") }}
```
Comment (VarunSharma15) [inline dbt/include/teradata/macros/adapters.sql:141]: In my opinion, it would be better if we can remove these logging statements.

File: dbt/include/teradata/macros/adapters.sql (hunk 2 above)
Comment (MT255026) [reply dbt/include/teradata/macros/adapters.sql:131]: Not required, will remove that

File: dbt/include/teradata/macros/adapters.sql (hunk 3 above)
Comment (MT255026) [reply dbt/include/teradata/macros/adapters.sql:141]: Sure, will remove that

**Review by VarunSharma15 — APPROVED** [review]
Looks Good To Me

File: README.md [hunk 4]
```diff
@@ -343,31 +343,6 @@ All dbt commands are supported.
 
 ### Custom configurations
 
-#### General
```
Comment (sc250072) [inline README.md:346]: image
This is not related to QVCI. Why are we mentioning about 1.8.x and again 1.8.2 in python compatibility and in dbt dependent package table, specified as1.8.x is compatible with 1.2.0 and 1.30 of dbt-teradata-util. Is this correct?

Comment (sc250072) [reply README.md:346]: At a minimum, you need to specify `host`, `user`, `password`, `schema` (database), `tmode`.

tmode is default to ANSI in https://github.com/Teradata/dbt-teradata/blob/3d42c353dc1e3fa3a08e64720a051ebd678149b6/dbt/adapters/teradata/connections.py#L28C4-L28C34 and its not mandatory. We can remove tmode from At a minimum section

File: dbt/include/teradata/macros/adapters.sql [hunk 5]
```diff
@@ -118,8 +118,23 @@
...
-    {% set use_qvci = var("use_qvci", "false") | as_bool %}
-     {{ log("use_qvci set to : " ~ use_qvci) }}
+    {% call statement('check_table_or_view', fetch_result=True) %}
+        SELECT TableKind FROM DBC.TablesV WHERE DatabaseName = '{{ relation.schema }}' AND TableName = '{{ relation.identifier }}'
+    {% endcall %}
+
+    {% set table_kind = load_result('check_table_or_view').table.columns['TableKind'].values()[0] %}
+
+    {%- if table_kind == 'V ' -%}
+        {% set temp_relation_for_view = relation.identifier ~ '_viw_tbl' %}
```
Comment (sc250072) [inline dbt/include/teradata/macros/adapters.sql:128]: can we use term `temp` label `temp_tbl` or 'temp_viw_tbl' to make it clear that its `temp`

File: dbt/include/teradata/macros/adapters.sql [hunk 6]
```diff
@@ -118,8 +118,23 @@
...
 {% macro teradata__get_columns_in_relation(relation) -%}
-    {% set use_qvci = var("use_qvci", "false") | as_bool %}
-     {{ log("use_qvci set to : " ~ use_qvci) }}
+    {% call statement('check_table_or_view', fetch_result=True) %}
+        SELECT TableKind FROM DBC.TablesV WHERE DatabaseName = '{{ relation.schema }}' AND TableName = '{{ relation.identifier }}'
+    {% endcall %}
+
+    {% set table_kind = load_result('check_table_or_view').table.columns['TableKind'].values()[0] %}
+
+    {%- if table_kind == 'V ' -%}
```
Comment (sc250072) [inline dbt/include/teradata/macros/adapters.sql:127]: Can we trim table_kind instead of checking equality with space after V

File: dbt/include/teradata/macros/adapters.sql [hunk 7]
```diff
@@ -178,31 +193,42 @@
       NULL AS table_database,
       ColumnsV.DatabaseName AS table_schema,
       ColumnsV.TableName AS table_name,
+      {%- if table_kind == 'V ' -%}
+        'view' AS table_type,
+      {%- else -%}
       CASE WHEN TablesV.TableKind = 'T' THEN 'table'
         WHEN TablesV.TableKind = 'O' THEN 'table'
         WHEN TablesV.TableKind = 'V' THEN 'view'
```
Comment (sc250072) [inline dbt/include/teradata/macros/adapters.sql:201]: Do we require TablesV.TableKind = 'V' THEN 'view' ?

File: dbt/include/teradata/macros/catalog.sql [hunk 8]
```diff
@@ -1,7 +1,49 @@
...
+            {%- endif -%}
+        {% endif %}
+    {%- endfor %}
+    {{ return(view_relations) }}
+{%- endmacro %}
+
+{% macro teradata__create_tmp_tables_of_views(view_relations) -%}
+    {% set view_tmp_tables_mapping = {} %}
+    {%- for relation in view_relations -%}
+        {% set temp_relation_for_view = relation.identifier ~ '_viw_tbl' %}
```
Comment (sc250072) [inline dbt/include/teradata/macros/catalog.sql:20]: as mentioned previously, please add temp label to temporary table name

File: tests/functional/adapter/test_unit_tests.py [hunk 9]
```diff
@@ -32,16 +32,6 @@
 
 class TestTestingTypesTeradata(BaseUnitTestingTypes):
 
-    @pytest.fixture(scope="class")
```
Comment (sc250072) [inline tests/functional/adapter/test_unit_tests.py:35]: Why these tests are removed

File: dbt/include/teradata/macros/catalog.sql [hunk 10]
```diff
@@ -175,13 +234,22 @@
...
-{% macro teradata__get_catalog_relations_where_clause_sql(relations) -%}
+{% macro teradata__get_catalog_relations_where_clause_sql(relations, view_tmp_tables_mapping) -%}
     where (
         {%- for relation in relations -%}
             {% if relation.schema and relation.identifier %}
                 (
                     upper("table_schema") = upper('{{ relation.schema }}')
-                    and upper("table_name") = upper('{{ relation.identifier }}')
+                    {% if view_tmp_tables_mapping is not none and view_tmp_tables_mapping|length > 0 %}
+                        {% set temp_table = view_tmp_tables_mapping[relation] %}
```
Comment (sc250072) [inline dbt/include/teradata/macros/catalog.sql:244]: If relation is added to view_tmp_tables_mapping means its view and mapping has happened with temp table. In this case, temp_table won't be empty. So checking of temp_table is only safe side but it won't be try at any time. Is this understanding correct?

**Review by sc250072 — CHANGES_REQUESTED** [review]
Couple of review comments added. Please address them. Rest all looks Good. Did we tested with all combinations of models like only tables, views and tables with views. And also, did you test with large number of models to check previous performance vs current performance?

File: README.md (hunk 4 above)
Comment (tallamohan) [reply README.md:346]: Yes, 1.8.x is compatible with 1.2.0 and 1.30 of dbt-teradata-utils

File: dbt/include/teradata/macros/adapters.sql (hunk 5 above)
Comment (tallamohan) [reply dbt/include/teradata/macros/adapters.sql:128]: Sure, I'll make this change and append "tmp_viw_tbl" because dbt-core uses the "tmp" annotation for temporary tables.

File: tests/functional/adapter/test_unit_tests.py (hunk 9 above)
Comment (tallamohan) [reply tests/functional/adapter/test_unit_tests.py:35]: We haven't removed those test cases; instead, we have overridden and customized them, changing the materialization to "table" because unit tests for views are not supported. With these changes, the unit test for the view now works as expected, and there is no need for any further overrides or changes. We are now running the unit tests as they are from dbt-core.

File: dbt/include/teradata/macros/catalog.sql (hunk 10 above)
Comment (tallamohan) [reply dbt/include/teradata/macros/catalog.sql:244]: This is required if any unit tests on a model that contains only tables.

File: dbt/include/teradata/macros/adapters.sql (hunk 7 above)
Comment (tallamohan) [reply dbt/include/teradata/macros/adapters.sql:201]: Yes, this is not required.

File: dbt/include/teradata/macros/catalog.sql (hunk 8 above)
Comment (tallamohan) [reply dbt/include/teradata/macros/catalog.sql:20]: Done, acknowledged above.

**Comment by tallamohan** [comment]
> Couple of review comments added. Please address them. Rest all looks Good. Did we tested with all combinations of models like only tables, views and tables with views. And also, did you test with large number of models to check previous performance vs current performance?

Here is the reference to the performance results:
* With code changes: 1817.69s (0:30:17)
* Without code changes: 1901.14s (0:31:41)

**Comment by tallamohan** [comment]
> Couple of review comments added. Please address them. Rest all looks Good. Did we tested with all combinations of models like only tables, views and tables with views. And also, did you test with large number of models to check previous performance vs current performance?

> Yes, I have tested with all combinations 
* All views
* Views and Tables
* All Tables

**Comment by sc250072** [comment]
> > Couple of review comments added. Please address them. Rest all looks Good. Did we tested with all combinations of models like only tables, views and tables with views. And also, did you test with large number of models to check previous performance vs current performance?
> 
> Here is the reference to the performance results:
> 
> * Without code changes: 1817.69s (0:30:17)
> * With code changes: 1901.14s (0:31:41)

Awesome!. There is not much difference. Whats the model size mohan for these results.

**Comment by sc250072** [comment]
Looks Good. Thank you @tallamohan . Good Work.

**Review by sc250072 — APPROVED** [review]
